src/curve.rs

```diff
@@ -0,0 +1,326 @@
+    // TODO: improve the constraint count by using an improved addition algorithm
+    pub fn add(
+        builder: &mut GadgetBuilder<F>,
+        point_1: &EdwardsPointExpression<F, C>,
+        point_2: &EdwardsPointExpression<F, C>,
+    ) -> EdwardsPointExpression<F, C> {
+        let d = C::d();
+        let a = C::a();
+        // TODO: better method for specifying variables
+        let EdwardsPointExpression { x: x1, y: y1, phantom } = point_1;
+        let EdwardsPointExpression { x: x2, y: y2, phantom } = point_2;
+        let x1y2 = builder.product(&x1, &y2);
+        let x2y1 = builder.product(&y1, &x2);
+        let x1x2 = builder.product(&x1, &x2);
+        let x1x2y1y2 = builder.product(&x1y2, &x2y1);
+        let y1y2 = builder.product(&y1, &y2);
+        let x3 = builder.quotient(&(x1y2 + x2y1), &(&x1x2y1y2 * &d + Expression::one()));
+        let y3 = builder.quotient(&(y1y2 - &x1x2 * &a), &(&x1x2y1y2 * -&d + Expression::one()));
+        EdwardsPointExpression::new_unsafe(x3, y3)
+    }
+
+    // TODO: improve constraint count
+    /// Naive implementation of the doubling algorithm for twisted Edwards curves.
+    ///
+    /// Assuming that EdwardsPointExpressions are on the curve, so the non-deterministic
+    /// division method is acceptable, as the denominator is non-zero.
+    ///
+    /// Note that this algorithm requires the point to be of odd order, which, in the case
+    /// of prime-order subgroups of Edwards curves, is satisfied.
+    pub fn double(
+        builder: &mut GadgetBuilder<F>,
+        point: &EdwardsPointExpression<F, C>,
+    ) -> EdwardsPointExpression<F, C> {
+        let EdwardsPointExpression { x, y, phantom } = point;
+        let a = C::a();
+
+        let xy = builder.product(&x, &y);
+        let xx = builder.product(&x, &x);
+        let yy = builder.product(&y, &y);
+        let x_2 = builder.quotient(&(&xy * Element::from(2u8)), &(&xx * &a + &yy));
+        let y_2 = builder.quotient(&(&yy - &xx * &a),
+                                   &(-&xx * &a - &yy + Expression::from(2u8)));
+
+        EdwardsPointExpression::new_unsafe(x_2, y_2)
+    }
+
+    /// Multiplies an EdwardsPointExpression by a scalar using a naive approach consisting of
+    /// multiplication by doubling.
+    // TODO: implement Daira's algorithm from https://github.com/zcash/zcash/issues/3924
+    pub fn scalar_mult(
+        builder: &mut GadgetBuilder<F>,
+        point: &EdwardsPointExpression<F, C>,
+        scalar: &Expression<F>
+    ) -> EdwardsPointExpression<F, C> {
+        let scalar_binary = builder.split_allowing_ambiguity(&scalar);
+
+        let mut sum = Self::identity();
+        let mut current = point;
+        for bit in scalar_binary.bits {
+            let boolean_product= &Self::boolean_mult(builder, current, &bit);
+            sum = Self::add(builder, &sum,boolean_product);
+            let current = &Self::double(builder, current);
```

Comment:
Might have been my mistake, but we should delete the `let` so that this mutates rather than making a new var.

Reply:
So this is a weird rust thing that I don't quite get - if we remove the let, it fails to compile and the compiler suggests adding let.

Reply:
Has this been removed or can we ignore the `let` being an issue here?

Reply:
Yes, Daniel helped me refactor, should be good now.